aip/auth/4115.md

```diff
@@ -7,33 +7,150 @@ created: 2020-08-13
+    configured with service account impersonation (the default), the identity
+    token endpoint always returns an error.
+  * When Workload Identity Federation for GKE is enabled, and the pod is
+    configured with service account impersonation, the identity token endpoint
+    returns a JWT as issued by
+    [generateIdToken](https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateIdToken)
+    for the Google service account being impersonated.
+
+### Retrieve and Cache Workload Access Tokens
+
+When retrieving workload access tokens from the metadata server on GCE, GKE, App
+Engine, or other GCP compute platforms, the following recommendations apply.
+Google-provided auth libraries adhere to these recommendations.  If your
+workload directly communicates with the metadata server in order to retrieve
+workload access tokens, you should adhere to these recommendations.  If you do
+not, your workload may suffer from intermittent and difficult-to-debug
+authentication errors.
+
+**Cache the access token in memory:** Your workload should not make a call to
+the metadata server every time you make a call to a GCP API.  Doing so will
+cause your workload to be rate-limited by the metadata server, or GCP IAM.
+Instead, you should maintain an in-memory cache of the access token, and use the
+cached token across all of your outbound requests.
+
+**Use a robust refresh strategy:**  Each time you attempt to use the cached
+access token, check the remaining lifetime of the token.
+* If it is greater than 225 seconds, the token is fresh.  Proceed to use it.
+* If it is less than 0 seconds, then the token is expired.  Refresh the cached
+  token, then proceed with the current thread.
+* If it is less than 225 seconds, but more than 0 seconds, then the token is
+  stale.  You can handle the current thread in one of two ways:
+  * (Background Refresh) If the token has greater than two minutes of validity
+    remaining, you can refresh the cached token in the background, allowing the
+    current thread to immediately proceed with the stale token, OR
+  * (Blocking Refresh) Refresh the cached token, blocking the current thread
+    until the refresh is complete.
+
+It is not safe to refresh the cached access token in the background on a
+schedule without additionally checking the status of the token before using it
+to make a request. Clock skew between your workload and the metadata server may
+result in your background refresh attempt still receiving a stale access token.
+
+Note that the access token returned by the metadata server may itself be stale.
+Certain implementations of the metadata server use the "Background Refresh"
+strategy described above for managing their own internal caches of tokens.  For
+example, when running on GKE with Workload Identity Federation for GKE enabled,
+gke-metadata-server will not reliably return a refreshed access token until 120
```

Comment:
If we will get back stale token until 2 mins before expiry, why should we try refresh 225 secs before expiry?